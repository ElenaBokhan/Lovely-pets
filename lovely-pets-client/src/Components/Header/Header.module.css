.header {
    display: flex;
    padding: 12px 0;
    align-items: center;
    gap: 80px;
    height: 80px;
    justify-content: space-between;
}

.profilesIcons {
    display: flex;
    width: 100px;
    justify-content: space-between;
}

.profilesIcon{
    display: flex;
}

.profilesIconImage {
    display: block;
}

.logo * {
    stroke: rgb(var(--text));
}

@media (width <= 576px) {
    .header {
        padding: 25px;
        position: relative;
        height: 170px;
        align-items: flex-start;
    }

    .profilesIcons {
        width: 120px;
        padding-top: 14px;
    }
}