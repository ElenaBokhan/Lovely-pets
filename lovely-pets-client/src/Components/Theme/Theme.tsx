import {createContext, useEffect, useState} from 'react';
import {THEME_STORAGE_KEY} from 'Const';

const supportedThemes = {
    light: 'light',
    dark: 'dark',
};

type Themes = keyof typeof supportedThemes;

export const ThemeContext = createContext<
    | {
          theme: Themes;
          setTheme: (theme: Themes) => void;
          supportedThemes: {[key: string]: string};
      }
    | undefined
>(undefined);

const getTheme = (): Themes => {
    let theme = localStorage.getItem(THEME_STORAGE_KEY);

    if (!theme) {
        localStorage.setItem(THEME_STORAGE_KEY, 'light');
        theme = 'light';
    }

    return theme as Themes;
};

const Theme = (props: {children: React.ReactNode}) => {
    const [theme, setTheme] = useState<Themes>(getTheme);

    useEffect(() => {
        localStorage.setItem(THEME_STORAGE_KEY, theme);
        document.documentElement.setAttribute('data-theme', theme);
    }, [theme]);

    return (
        <ThemeContext.Provider
            value={{
                theme,
                setTheme,
                supportedThemes,
            }}
        >
            {props.children}
        </ThemeContext.Provider>
    );
};

export default Theme;
