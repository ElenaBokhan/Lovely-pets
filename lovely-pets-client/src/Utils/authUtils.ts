import {ILoginResponse} from 'Api/authClient';

export const setLoginToStorage = (loginData: ILoginResponse) => {
    const {access_token, refresh_token, username, id} = loginData;

    localStorage.setItem('access_token', access_token);
    localStorage.setItem('refresh_token', refresh_token);
    localStorage.setItem('userName', username);
    localStorage.setItem('userId', id);
};

export const removeLoginToStorage = () => {
    localStorage.removeItem('access_token');
    localStorage.removeItem('refresh_token');
    localStorage.removeItem('userName');
    localStorage.removeItem('userId');
};
