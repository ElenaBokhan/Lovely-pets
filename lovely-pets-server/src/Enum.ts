export enum ESortFilter {
    POPULAR = 'POPULAR',
    NEW = 'NEW',
    CHEAPER = 'CHEAPER',
    EXPENSIVE = 'EXPENSIVE',
    RATING = 'RATING',
    DISCOUNT = 'DISCOUNT',
}
