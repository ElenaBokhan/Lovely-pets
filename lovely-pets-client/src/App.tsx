import './index.css';
import {Layout} from 'Components/Layout/Layout';
import {Catalog} from 'Pages/Catalog/Catalog';
import {Favourites} from 'Pages/Favourites/Favourites';
import {PageNotFound} from 'Pages/PageNotFound/PageNotFound';
import {PetCard} from 'Pages/PetCard/PetCard';
import {Profile} from 'Pages/Profile/Profile';
import {RouterProvider, createBrowserRouter, RouteObject} from 'react-router-dom';
import {ToastContainer} from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import {SignUp} from 'Pages/SignUp/SignUp';
import {SignIn} from 'Pages/SignIn/SignIn';
import {Cart} from 'Pages/Cart/Cart';
import {AddReview} from 'Pages/AddReview/AddReview';
import {AddProduct} from 'Pages/AddProduct/AddProduct';
import {QueryClient, QueryClientProvider} from '@tanstack/react-query';
import {withTranslation} from 'react-i18next';
import {useEffect} from 'react';
import {setCartFromStorage, setCartToStorage} from 'Store/cartStore';
import 'Style/Themes.module.css';
import Theme from 'Components/Theme/Theme';

export const router: RouteObject[] = [
    {
        path: '/',
        element: <Layout />,
        errorElement: <div />, // TODO: добавить errorBoundary, пока заглушка
        children: [
            {
                path: '/',
                element: <Catalog />,
            },
            {
                path: '/favourites',
                element: <Favourites />,
            },
            {
                path: '/pets/:petId',
                element: <PetCard />,
            },
            {
                path: '/profile',
                element: <Profile />,
            },
            {
                path: '/addReview',
                element: <AddReview />,
            },
            {
                path: '/addPet',
                element: <AddProduct />,
            },
            {
                path: '/signup',
                element: <SignUp />,
            },
            {
                path: '/signin',
                element: <SignIn />,
            },
            {
                path: '/cart',
                element: <Cart />,
            },
            {
                path: '*',
                element: <PageNotFound />,
            },
        ],
    },
];

export const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            retry: false,
        },
    },
});

const App = () => {
    useEffect(() => {
        window.addEventListener('load', setCartFromStorage);
        window.addEventListener('beforeunload', setCartToStorage);
        return () => {
            window.removeEventListener('load', setCartFromStorage);
            window.removeEventListener('beforeunload', setCartToStorage);
        };
    }, []);

    return (
        <QueryClientProvider client={queryClient}>
            <Theme>
                <RouterProvider router={createBrowserRouter(router)} />
                <ToastContainer autoClose={2000} position="top-right" theme="light" closeOnClick />
            </Theme>
        </QueryClientProvider>
    );
};

export default withTranslation()(App);
