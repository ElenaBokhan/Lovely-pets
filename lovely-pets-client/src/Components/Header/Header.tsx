import {ReactComponent as LogOut} from 'assets/log-out.svg';
import {ReactComponent as AddNewIcon} from 'assets/add-new.svg';
import HeaderContainer, {EContainerType} from 'Components/Common/Container/Container';
import styles from 'Components/Header/Header.module.css';
import {SearchForm} from 'Components/SearchForm/SearchForm';
import {useLocation, useNavigate} from 'react-router-dom';
import {ETestId} from 'Enum';
import {removeLoginToStorage} from 'Utils/authUtils';
import {ThemeToggler} from 'Components/ThemeToggler/ThemeToggler';
import {Logo} from 'Components/Common/Logo/Logo';
import {LocaleToggler} from 'Components/LocaleToggler/LocaleToggler';

export const Header = () => {
    const navigate = useNavigate();
    const {pathname} = useLocation();

    const redirect = (path: string) => () => navigate(path);

    const handleLogOut = () => {
        removeLoginToStorage();
        navigate('/signin');
    };

    const renderHeaderProfileIcons = () => {
        return (
            <div className={styles.profilesIcons}>
                <LocaleToggler />
                <ThemeToggler />
                <AddNewIcon
                    className={styles.logo}
                    onClick={redirect('/addPet')}
                    data-test-id={ETestId.HEADER_ADD_NEW_ICON}
                />
                <LogOut className={styles.logo} onClick={handleLogOut} data-test-id={ETestId.HEADER_LOG_OUT_ICON} />
            </div>
        );
    };

    return (
        <HeaderContainer className={styles.header} type={EContainerType.HEADER}>
            <Logo />
            {pathname === '/' && <SearchForm />}
            {!['/signin', '/signup'].includes(pathname) && renderHeaderProfileIcons()}
        </HeaderContainer>
    );
};
