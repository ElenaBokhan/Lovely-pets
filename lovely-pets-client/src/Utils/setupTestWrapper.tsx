import {queryClient, router} from 'App';
import {render} from '@testing-library/react';
import {createMemoryRouter, RouterProvider} from 'react-router-dom';
import {ToastContainer} from 'react-toastify';
import {QueryClientProvider} from '@tanstack/react-query';
import Theme from 'Components/Theme/Theme';

interface WrapperForTestsParams {
    initialRouterEntries: string[];
}

export const setupTestsWrapper = ({initialRouterEntries}: WrapperForTestsParams) => {
    const memoryRouter = createMemoryRouter(router, {
        initialEntries: initialRouterEntries,
    });
    render(
        <QueryClientProvider client={queryClient}>
            <Theme>
                <RouterProvider router={memoryRouter} />
                <ToastContainer autoClose={2000} position="top-right" theme="light" closeOnClick />
            </Theme>
        </QueryClientProvider>
    );

    return memoryRouter;
};
