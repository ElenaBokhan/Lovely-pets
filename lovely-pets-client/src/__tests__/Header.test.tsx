import '@testing-library/jest-dom';
import {setupTestsWrapper} from 'Utils/setupTestWrapper';
import {getByTestId, render, screen} from '@testing-library/react';
import {ETestId} from 'Enum';
import Theme from 'Components/Theme/Theme';
import {QueryClientProvider} from '@tanstack/react-query';
import {queryClient} from 'App';
import {BrowserRouter} from 'react-router-dom';
import {Header} from 'Components/Header/Header';

describe('Тесты для компонента Header', () => {
    beforeEach(() => {
        setupTestsWrapper({initialRouterEntries: ['/']});
    });

    describe('Проверяем доступность элементов под неавторизованным пользователем', () => {
        checkHeaderMainLogoIsAvailable();
        checkHeaderProfilesIconAvailable(false);
        checkHeaderSearchInputAvailable(false);
    });

    describe('Проверяем доступность ссылок под авторизованным пользователем', () => {
        beforeEach(() => {
            jest.spyOn(Storage.prototype, 'getItem').mockImplementation(() => 'token');
        });

        checkHeaderMainLogoIsAvailable();
        checkHeaderProfilesIconAvailable(true);
        checkHeaderSearchInputAvailable(true);
    });

    describe('Снапшот компонента Header', () => {
        test('Snapshot', () => {
            const {asFragment} = render(
                <QueryClientProvider client={queryClient}>
                    <Theme>
                        <BrowserRouter basename="/">
                            <Header />
                        </BrowserRouter>
                    </Theme>
                </QueryClientProvider>
            );

            expect(asFragment()).toMatchSnapshot();
        });
    });
});

function checkHeaderMainLogoIsAvailable() {
    test('Проверяем доступность логотипа', () => {
        const header = screen.getByTestId(ETestId.HEADER);
        const headerLogo = getByTestId(header, ETestId.MAIN_LOGO);
        expect(headerLogo).toBeInTheDocument();
    });
}

function checkHeaderProfilesIconAvailable(enable: boolean) {
    if (enable) {
        test('Проверяем доступность иконок профиля', () => {
            expect(screen.getByTestId(ETestId.HEADER_LOCALE_TOGGLE)).toBeInTheDocument();
            expect(screen.getByTestId(ETestId.HEADER_THEME_TOGGLE)).toBeInTheDocument();
            expect(screen.getByTestId(ETestId.HEADER_ADD_NEW_ICON)).toBeInTheDocument();
            expect(screen.getByTestId(ETestId.HEADER_LOG_OUT_ICON)).toBeInTheDocument();
        });
    } else {
        test('Проверяем не доступность иконок профиля', () => {
            expect(screen.queryByTestId(ETestId.HEADER_LOCALE_TOGGLE)).not.toBeInTheDocument();
            expect(screen.queryByTestId(ETestId.HEADER_THEME_TOGGLE)).not.toBeInTheDocument();
            expect(screen.queryByTestId(ETestId.HEADER_ADD_NEW_ICON)).not.toBeInTheDocument();
            expect(screen.queryByTestId(ETestId.HEADER_LOG_OUT_ICON)).not.toBeInTheDocument();
        });
    }
}

function checkHeaderSearchInputAvailable(enable: boolean) {
    if (enable) {
        test('Проверяем доступность компонента поиска', () => {
            expect(screen.getByTestId(ETestId.HEADER_SEARCH_INPUT)).toBeInTheDocument();
        });
    } else {
        test('Проверяем не доступность компонента поиска', () => {
            expect(screen.queryByTestId(ETestId.HEADER_SEARCH_INPUT)).not.toBeInTheDocument();
        });
    }
}
