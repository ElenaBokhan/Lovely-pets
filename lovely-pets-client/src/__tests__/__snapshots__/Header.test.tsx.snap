// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тесты для компонента Header Снапшот компонента Header Snapshot 1`] = `
<DocumentFragment>
  <header
    class="headerContainer"
    data-test-id="header"
  >
    <div
      class="container header"
    >
      <div
        class="logoWrapper"
      >
        <div
          alt="logoMain"
          class="logo"
          data-test-id="mainLogo"
          height="56"
        />
        <span
          class="undefined bold head1"
        >
          LovelyPets
        </span>
      </div>
      <div
        class="searchForm"
      >
        <form>
          <input
            data-test-id="headerSearchInput"
            placeholder="Поиск товаров"
            value=""
          />
        </form>
        <button
          class="button"
          type="button"
        >
          <img
            alt="searchButton"
            src="SvgrURL"
          />
        </button>
      </div>
      <div
        class="profilesIcons"
      >
        <button
          class="locale"
          data-test-id="headerLocaleToggle"
        >
          EN
        </button>
        <div
          class="logo"
          data-test-id="headerThemeToggle"
        />
        <div
          class="logo"
          data-test-id="headerAddNewProduct"
        />
        <div
          class="logo"
          data-test-id="headerLogOutIcon"
        />
      </div>
    </div>
  </header>
</DocumentFragment>
`;
